Pcap++/header/KniDevice.h

```diff
@@ -0,0 +1,709 @@
+		 * This structure MUST be used ONLY when KniDevice#callbackVersion returns
+		 * KniCallbackVersion#CALLBACKS_NEW.
+		 * Or if You are sure that DPDK version used is 17.11 or higher.
+		 * If some callback is not provided (NULL) the request will always succeeds
+		 * if other is not specified in callback description.
+		 */
+		struct KniIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint16_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint16_t port_id, uint8_t if_up);
+			/**
+			 * Pointer to function of configuring mac address.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets MAC address of
+			 * DPDK port port_id via rte_eth_dev_default_mac_addr_set.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_mac_address)(uint16_t port_id, uint8_t mac_addr[]);
+			/**
+			 * Pointer to function of configuring promiscuous mode.
+			 * If callback is not provided and port_id of KNI device is not UINT16_MAX
+			 * then DPDK will use predefined callback witch sets promiscuous mode of
+			 * DPDK port port_id via rte_eth_promiscuous_enable/rte_eth_promiscuous_disable.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_promiscusity)(uint16_t port_id, uint8_t to_on);
+		};
+
+		/**
+		 * @brief Old callbacks for KNI device events.
+		 * This structure MUST be used ONLY when KniDevice#callbackVersion returns
+		 * KniCallbackVersion#CALLBACKS_OLD.
+		 * Or if You are sure that DPDK version used is lower than 17.11.
+		 * If some callback is not provided (NULL) the request will always succeeds.
+		 */
+		struct KniOldIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint8_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint8_t port_id, uint8_t if_up);
+		};
+
+		/**
+		 * @brief KNI device initialization data.
+		 * Used to create new KNI device.
+		 * Usage of callbacks member or oldCallbacks member is defined by
+		 * result of KniDevice#callbackVersion
+		 */
+		struct KniDeviceConfiguration
+		{
+			enum
+			{
+				// Must correspond to minimum of
+				// RTE_KNI_NAMESIZE [DPDK] = 32 and
+				// IFNAMSIZ [LINUX] = 16 (on most systems)
+				KNI_NAME_SIZE = 16
+			};
+			/**
+			 * Name used to display device in system.
+			 * Must not interfere with already existing network interfaces.
+			 */
+			char name[KNI_NAME_SIZE];
```

Comment:
I prefer for this structure to not to have any overhead that comes with std::string (like it is not a POD type so the rule of 3 or rule of 5 then must apply and i don't want to write this usless constructors and assignment operators especialy in C++98 where there is no `= default` expression they will overweight the code base for no good reason). This is a simple C structure and it must stay this way.
The better reason for this is that there is no way to express the string of fixed length in C++. Only the C char array will have this property

Reply:
I agree but still from the user point of view this is not convenient. It's much easier to work with `std::string` than with char arrays. If you use `std::string` that doesn't mean you need to create c'tors and d'tors for `KniDeviceConfiguration` because `std::string` already has default c'tor, copy c'tor and so on.

I do agree on the other reason you've mentioned, but it's a small price to pay for a more easy-to-use C++ API (which is what PcapPlusPlus is all about)

Reply:
See: https://github.com/seladb/PcapPlusPlus/pull/196#discussion_r286646031

Reply:
\+